internal/introspect/invokes.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package introspect
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// Retrieve the Input type of a function, if any.
+func InvokeInput(f reflect.Type) (input reflect.Type, hasContext bool, err error) {
+	contract.Assert(f.Kind() == reflect.Func)
+	isContext := func(t reflect.Type) bool { return t.Implements(reflect.TypeOf(new(context.Context)).Elem()) }
+	badTypeMsg := fmt.Errorf("Functions must be of the type func([context.Context, ], T). Found type %s", f.String())
+	switch f.NumIn() {
```

Comment:
I don't understand why we limit the signature of a function so strictly. I see that most/all of our functions on the registry have some args struct as input and a result struct as an output, could we not automatically wrap the the inputs and outputs into a struct?

Reply:
The schema (and the gRPC protocol) have functions pass input structs and return output structs, so we need to end up with that. We can't do automatic wrapping of multiple inputs and outputs into their struct components because reflection does not provide us with argument names, so we won't know what keys to provide in the newly created structs. 

Given:
```go
func AddWithRemainder(x, y int) (sum int, remainder int) {
    return (int)((u8)x+y), (x+y)%(2^8) 
}
```
We can only retrieve
```go
reflect.TypeOf(AddWithRemainder).String() == "func(int, int) (int, int)"
```